README.md
What is TNAT64?
===============

TNAT64 is an interceptor which redirects outgoing TCPv4 connections
through NAT64, thus enabling an application running on an IPv6-only host
to communicate with the IPv4 world, even if that application does not
support IPv6 at all.

TNAT64 is based on the original code of [tsocks].

What is NAT64?
==============

According to [Wikipædia], '*NAT64 is a mechanism to allow IPv6 hosts to
communicate with IPv4 servers. The NAT64 server is the endpoint for at
least one IPv4 address and an IPv6 network segment of 32-bits (e.g.
64:FF9B::/96). The IPv6 client embeds the IPv4 address it wishes to
communicate with using these bits, and sends its packets to the
resulting address. The NAT64 server then creates a NAT-mapping between
the IPv6 and the IPv4 address, allowing them to communicate.*'

How does it work?
=================

From the user’s point of view, the application is run by prepending its
name with *tnat64*:

    $ tnat64 some-old-application

And then magic happens: when on an IPv6-only network, the application
uses NAT64 facilities to contact IPv4 hosts, and when IPv6 is not
available, but IPv4 is functional, IPv4 is used. Everything happens
without the need to reconfigure or recompile anything.

But how does it work deep inside?
=================================

TNAT64 is implemented as a shared object library which is loaded into
your application by means of the [ld.so(8)] `LD_PRELOAD` feature. The
library intercepts certain BSD socket calls (namely, [socket(2)] and
[connect(2)]). When an application creates a socket of `AF_INET` family
and `SOCK_STREAM` type, that is, a TCPv4 socket, that call is
intercepted, and an `AF_INET6` socket is returned instead. Following
this, any `connect(2)` call on this socket is also intercepted.
TNAT64 then decides whether the destination host is reachable over IPv4,
or instead needs to be forwarded to the NAT64.

The decision is made as follows:

1.  First, the configuration file is consulted to find the appropriate
    NAT64 prefix. If the prefix isn\'t specified for the destination
    network, this issue is reported to the user.
2.  The NAT64 destination address is constructed by appending the IPv4
    destination address to the NAT64 prefix. E.g. the destination
    address 198.51.100.192 and the NAT64 prefix [64:ff9b::/96] would
    result in the NAT64 address 64:ff9b::198.51.100.192 (aka
    64:ff9b::c933:64c0).
3.  A connection attempt is performed to that address.
4.  If the connection attempt fails because the destination network is
    unreachable, the IPv4-mapped IPv6 address is constructed and a
    connection attempt to that address is performed. These addresses
    have the prefix ::ffff:0:0/96 and are constructed just as described
    above. The connect call to this address is interpreted by the
    network stack the same way as if an TCPv4 connection was requested,
    thus the connection is made over IPv4 and not IPv6. This way the
    host can be reached via IPv4 when IPv6 networking is not available.
5.  The next time the connection is requested, the host is contacted
    using the last successful method.

[tsocks]: http://tsocks.sourceforge.net/
[Wikipædia]: http://en.wikipedia.org/wiki/NAT64
[ld.so(8)]: http://linux.die.net/man/8/ld.so
[socket(2)]: http://linux.die.net/man/2/socket
[connect(2)]: http://linux.die.net/man/2/connect
[64:ff9b::/96]: http://tools.ietf.org/html/rfc6052


ChangeLog
2021-10-13  Andrej Shadura    <andrew@shadura.me>

        * Version 0.06 released.
        * Use strerror() instead of sys_errlist to fix a build failure
          with glibc 2.32 (patch by Logan Rosen).
        * Update and reformat INSTALL instructions.
        * Add a manpage for tnat64-validateconf.

2011-09-16  Andrej Shadura    <andrew@shadura.me>

        * Version 0.05 released.

2011-09-16  Barak A. Pearlmutter  <barak@cs.nuim.ie>

        * Fix malloc return code checking error
        * Squash unused variable warnings
        * Autotools modernisation
          - Fix autoheader issue (crashed, did not generate config.h.in).
          - Create Makefile.am, use automake, rm manually-created Makefile.in.
          - Put m4 cache in m4/, make stub file there to preserve directory.
          - Generate tnat64 / tnat64.8 from tnat64.in / tnat64.8.in, thereby
            allowing them to be specialized to current configuration.
          - Delete files which can be generated by autoreconf --install
        * Update INSTALL instructions
        * Rename validateconf to tnet64-validateconf and include in installation
        * Add and slightly improve README.creole

2011-09-11  Andrej Shadura  <andrew@shadura.me>

        * Version 0.04 released.
        * The dynamic library moved into subdirectory.
        * Some minor fixes.

2011-05-19  Andrej Shadura  <andrew@shadura.me>

        * Version 0.03 released.
        * X11 compatibility (get*name functions now work properly).
        * Crash on wrong function calls order fixed.
        * Better debug facilities.

2011-04-29  Andrej Shadura  <andrew@shadura.me>

        * Version 0.02 released.
        * getsockname(2) and getpeername(2) are now intercepted, too;
          that improves the compatibility with some software.
        * Some bugs fixed.

2011-04-19  Andrej Shadura  <andrew@shadura.me>

        * The first release: 0.01.

2011-03-02  Andrej Shadura  <andrew@shadura.me>

        * Got it working!

2011-03-01  Andrej Shadura  <andrew@shadura.me>

        * First attempt to implement NAT64 wrapper,
          tsocks code is used as a starting point.


common.c
/*

    commmon.c    - Common routines for the tnat64 package 

*/

#include <config.h>
#include <stdio.h>
#include <netdb.h>
#include <common.h>
#include <stdarg.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/in.h>

/* Globals */
int loglevel = MSGERR;          /* The default logging level is to only log
                                   error messages */
char *logfilename = NULL;       /* Name of file to which log messages should
                                   be redirected */
FILE *logfile = NULL;           /* File to which messages should be logged */
int logstamp = 0;               /* Timestamp (and pid stamp) messages */

unsigned int HIDDENSYM resolve_ip(char *host, int showmsg, int allownames)
{
    struct hostent *new;
    unsigned int hostaddr;
    struct in_addr *ip;

    if ((hostaddr = inet_addr(host)) == (unsigned int)-1)
    {
        /* We couldn't convert it as a numerical ip so */
        /* try it as a dns name                        */
        if (allownames)
        {
#ifdef HAVE_GETHOSTBYNAME
            if ((new = gethostbyname(host)) == (struct hostent *)0)
            {
#endif
                return (-1);
#ifdef HAVE_GETHOSTBYNAME
            }
            else
            {
                ip = ((struct in_addr *)*new->h_addr_list);
                hostaddr = ip->s_addr;
                if (showmsg)
                    printf("Connecting to %s...\n", inet_ntoa(*ip));
            }
#endif
        }
        else
            return (-1);
    }

    return (hostaddr);
}

/* Set logging options, the options are as follows:             */
/*  level - This sets the logging threshold, messages with      */
/*          a higher level (i.e lower importance) will not be   */
/*          output. For example, if the threshold is set to     */
/*          MSGWARN a call to log a message of level MSGDEBUG   */
/*          would be ignored. This can be set to -1 to disable  */
/*          messages entirely                                   */
/*  filename - This is a filename to which the messages should  */
/*             be logged instead of to standard error           */
/*  timestamp - This indicates that messages should be prefixed */
/*              with timestamps (and the process id)            */
void HIDDENSYM set_log_options(int level, char *filename, int timestamp)
{

    loglevel = level;
    if (loglevel < MSGERR)
        loglevel = MSGNONE;

    if (filename)
    {
        logfilename = strdup(filename);
    }

    logstamp = timestamp;
}

void HIDDENSYM show_msg(int level, char *fmt, ...)
{
    va_list ap;
    int saveerr;
    extern char *progname;
    char timestring[20];
    time_t timestamp;

    if ((loglevel == MSGNONE) || (level > loglevel))
        return;

    if (!logfile)
    {
        if (logfilename)
        {
            logfile = fopen(logfilename, "a");
            if (logfile == NULL)
            {
                logfile = stderr;
                show_msg(MSGERR, "Could not open log file, %s, %s\n", logfilename, strerror(errno));
            }
        }
        else
            logfile = stderr;
    }

    if (logstamp)
    {
        timestamp = time(NULL);
        strftime(timestring, sizeof(timestring), "%H:%M:%S", localtime(&timestamp));
        fprintf(logfile, "%s ", timestring);
    }

    fputs(progname, logfile);

    if (logstamp)
    {
        fprintf(logfile, "(%d)", getpid());
    }

    fputs(": ", logfile);

    va_start(ap, fmt);

    /* Save errno */
    saveerr = errno;

    vfprintf(logfile, fmt, ap);

    fflush(logfile);

    errno = saveerr;

    va_end(ap);
}


parser.h
/* parser.h - Structures, functions and global variables for the */
/* tsocks parsing routines                                       */

#ifndef _PARSER_H

#define _PARSER_H	1

/* Structure definitions */

/* Structure representing one NAT64 prefix specified in the config */
struct prefixent
{
    int lineno;                 /* Line number in conf file this path started on */
    char *address;              /* IPv6 address prefix in textual form */
    struct in6_addr prefix;     /* the same, but in binary form */
    struct netent *reachnets;   /* Linked list of nets from this preifx */
    struct prefixent *next;     /* Pointer to next prefix entry */
};

/* Structure representing a network */
struct netent
{
    struct in_addr localip;     /* Base IP of the network */
    struct in_addr localnet;    /* Mask for the network */
    unsigned long startport;    /* Range of ports for the */
    unsigned long endport;      /* network                */
    struct netent *next;        /* Pointer to next network entry */
};

/* Structure representing a complete parsed file */
struct parsedfile
{
    struct netent *localnets;
    struct prefixent defaultprefix;
    struct prefixent *paths;
};

/* Functions provided by parser module */
int read_config(char *, struct parsedfile *);
int is_local(struct parsedfile *, struct in_addr *);
int pick_prefix(struct parsedfile *, struct prefixent **, struct in_addr *, uint16_t port);
int check_prefix(struct parsedfile *config, struct in6_addr * addr);
char *strsplit(char *separator, char **text, const char *search);

#endif


tsocks.h
/* tsocks.h - Structures used by tsocks to form SOCKS requests */

#ifndef _TSOCKS_H

#define _TSOCKS_H	1

#include <parser.h>

/* Structure representing a socks connection request */
struct sockreq
{
    int8_t version;
    int8_t command;
    int16_t dstport;
    int32_t dstip;
    /* A null terminated username goes here */
};

/* Structure representing a socks connection request response */
struct sockrep
{
    int8_t version;
    int8_t result;
    int16_t ignore1;
    int32_t ignore2;
};

/* Structure representing a socket which we are currently proxying */
struct connreq
{
    /* Information about the socket and target */
    int sockid;
    struct sockaddr_in connaddr;
    struct sockaddr_in serveraddr;

    /* Pointer to the config entry for the socks server */
    struct serverent *path;

    /* Current state of this proxied socket */
    int state;

    /* Next state to go to when the send or receive is finished */
    int nextstate;

    /* When connections fail but an error number cannot be reported 
     * because the socket is non blocking we keep the connreq struct until
     * the status is queried with connect() again, we then return
     * this value */
    int err;

    /* Events that were set for this socket upon call to select() or
     * poll() */
    int selectevents;

    /* Buffer for sending and receiving on the socket */
    int datalen;
    int datadone;
    char buffer[1024];

    struct connreq *next;
};

/* Connection statuses */
#define UNSTARTED 0
#define CONNECTING 1
#define CONNECTED 2
#define SENDING 3
#define RECEIVING 4
#define SENTV4REQ 5
#define GOTV4REQ 6
#define SENTV5METHOD 7
#define GOTV5METHOD 8
#define SENTV5AUTH 9
#define GOTV5AUTH 10
#define SENTV5CONNECT 11
#define GOTV5CONNECT 12
#define DONE 13
#define FAILED 14

/* Flags to indicate what events a socket was select()ed for */
#define READ (1<<0)
#define WRITE (1<<1)
#define EXCEPT (1<<2)
#define READWRITE (READ|WRITE)
#define READWRITEEXCEPT (READ|WRITE|EXCEPT)

#endif


validateconf.c
/*

    VALIDATECONF - Part of the tnat64 package
		   This utility can be used to validate the tnat64.conf
		   configuration file

    Copyright (C) 2000 Shaun Clowes 

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
    MA 02110-1301, USA.

*/

/* Global configuration variables */
char *progname = "tnat64-validateconf";        /* Name for error msgs      */

/* Header Files */
#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <string.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <common.h>
#include <parser.h>

void show_prefix(struct parsedfile *, struct prefixent *, int);
void show_conf(struct parsedfile *config);
void test_host(struct parsedfile *config, char *);

int main(int argc, char *argv[])
{
    char *usage = "Usage: [-f conf file] [-t hostname/ip[:port]]";
    char *filename = NULL;
    char *testhost = NULL;
    struct parsedfile config;
    int i;

    if ((argc > 5) || (((argc - 1) % 2) != 0))
    {
        show_msg(MSGERR, "Invalid number of arguments\n");
        show_msg(MSGERR, "%s\n", usage);
        exit(1);
    }

    for (i = 1; i < argc; i = i + 2)
    {
        if (!strcmp(argv[i], "-f"))
        {
            filename = argv[(i + 1)];
        }
        else if (!strcmp(argv[i], "-t"))
        {
            testhost = argv[(i + 1)];
        }
        else
        {
            show_msg(MSGERR, "Unknown option %s\n", argv[i]);
            show_msg(MSGERR, "%s\n", usage);
            exit(1);
        }
    }

    if (!filename)
        filename = strdup(CONF_FILE);

    printf("Reading configuration file %s...\n", filename);
    if (read_config(filename, &config) == 0)
        printf("... Read complete\n\n");
    else
        exit(1);

    /* If they specified a test host, test it, otherwise */
    /* dump the configuration                            */
    if (!testhost)
        show_conf(&config);
    else
        test_host(&config, testhost);

    return (0);
}

void test_host(struct parsedfile *config, char *host)
{
    struct in_addr hostaddr;
    struct prefixent *path;
    char *hostname, *port;
    char separator;
    unsigned long portno = 0;

    /* See if a port has been specified */
    hostname = strsplit(&separator, &host, ": \t\n");
    if (separator == ':')
    {
        port = strsplit(NULL, &host, " \t\n");
        if (port)
            portno = strtol(port, NULL, 0);
    }

    /* First resolve the host to an ip */
    if ((hostaddr.s_addr = resolve_ip(hostname, 0, 1)) == -1)
    {
        fprintf(stderr, "Error: Cannot resolve %s\n", host);
        return;
    }
    else
    {
        printf("Finding path for %s...\n", inet_ntoa(hostaddr));
        if (!(is_local(config, &(hostaddr))))
        {
            printf("Path is local\n");
        }
        else
        {
            pick_prefix(config, &path, &hostaddr, portno);
            if (path == &(config->defaultprefix))
            {
                printf("Path is reached via default NAT64 prefix:\n");
                show_prefix(config, path, 1);
            }
            else
            {
                printf("Host is reached via this path:\n");
                show_prefix(config, path, 0);
            }
        }
    }

    return;
}

void show_conf(struct parsedfile *config)
{
    struct netent *net;
    struct prefixent *prefix;

    /* Show the local networks */
    printf("=== Local networks (no NAT64 needed) ===\n");
    net = (config->localnets);
    while (net != NULL)
    {
        printf("Network: %15s ", inet_ntoa(net->localip));
        printf("Netmask: %15s\n", inet_ntoa(net->localnet));
        net = net->next;
    }
    printf("\n");

    /* If we have a default prefix configuration show it */
    printf("=== Default NAT64 prefix configuration ===\n");
    if ((config->defaultprefix).address != NULL)
    {
        show_prefix(config, &(config->defaultprefix), 1);
    }
    else
    {
        printf("No default NAT64 prefix specified, this is rarely a " "good idea\n");
    }
    printf("\n");

    /* Now show paths */
    if ((config->paths) != NULL)
    {
        prefix = (config->paths);
        while (prefix != NULL)
        {
            printf("=== Path (line no %d in configuration file)" " ===\n", prefix->lineno);
            show_prefix(config, prefix, 0);
            printf("\n");
            prefix = prefix->next;
        }
    }

    return;
}

void show_prefix(struct parsedfile *config, struct prefixent *prefix, int def)
{
    struct netent *net;

    /* Show address */
    if (prefix->address != NULL)
        printf("NAT64 prefix:       %s\n", prefix->address);
    else
        printf("NAT64 prefix:       ERROR! None specified\n");


    /* If this is the default servers and it has reachnets, thats stupid */
    if (def)
    {
        if (prefix->reachnets != NULL)
        {
            fprintf(stderr, "Error: The default NAT64 prefix has "
                    "specified networks it can be used to reach (subnet statements), "
                    "these statements are ignored since the " "default NAT64 prefix will be used for any network " "which is not specified in a subnet statement " "for other prefixes\n");
        }
    }
    else if (prefix->reachnets == NULL)
    {
        fprintf(stderr, "Error: No subnet statements specified for " "this NAT64 prefix, it will never be used\n");
    }
    else
    {
        printf("This NAT64 prefix can be used to reach:\n");
        net = prefix->reachnets;
        while (net != NULL)
        {
            printf("Network: %15s ", inet_ntoa(net->localip));
            printf("Netmask: %15s ", inet_ntoa(net->localnet));
            if (net->startport)
                printf("Ports: %5lu - %5lu", net->startport, net->endport);
            printf("\n");
            net = net->next;
        }
    }
}


tnat64.c
/*

    TNAT64 - Wrapper library for redirecting IPv4 connections to NAT64
    Copyright (C) 2011 Andrej Shadura
    Based on original code of TSOCKS, copyright (C) 2000 Shaun Clowes

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
    MA 02110-1301, USA.

*/

/* PreProcessor Defines */
#include <config.h>

#ifdef USE_GNU_SOURCE
#define _GNU_SOURCE
#endif

/* Global configuration variables */
char *progname = "libtnat64";   /* Name used in err msgs    */

/* Header Files */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <dlfcn.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/poll.h>
#include <string.h>
#include <strings.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/poll.h>
#include <sys/time.h>
#include <pwd.h>
#include <errno.h>
#include <fcntl.h>
#include <common.h>
#include <stdarg.h>
#include <parser.h>

/* Mask which covers at least up to SOCK_MASK-1.  The
 *  * remaining bits are used as flags.
 *  per Linux kernel 2.6.* */
#define SOCK_TYPE_MASK 0xf

/* Global Declarations */
static int (*realsocket) (SOCKET_SIGNATURE);
static int (*realconnect) (CONNECT_SIGNATURE);
static int (*realgetpeername) (GETPEERNAME_SIGNATURE);
static int (*realgetsockname) (GETSOCKNAME_SIGNATURE);
static struct parsedfile *config;
static int suid = 0;
static char *conffile = NULL;

static struct in6_addr ipv4mapped;

static int current_af = AF_INET6;

/* Exported Function Prototypes */
void _init(void);
int socket(SOCKET_SIGNATURE);
int connect(CONNECT_SIGNATURE);
int getpeername(GETPEERNAME_SIGNATURE);
int getsockname(GETSOCKNAME_SIGNATURE);

/* Private Function Prototypes */
static int get_config();
static int get_environment();

void _init(void)
{
#ifdef USE_OLD_DLSYM
    void *lib;
#endif

    /* We could do all our initialization here, but to be honest */
    /* most programs that are run won't use our services, so     */
    /* we do our general initialization on first call            */

    /* Determine the logging level */
    suid = (getuid() != geteuid());

#ifndef USE_OLD_DLSYM
    realconnect = dlsym(RTLD_NEXT, "connect");
    realsocket = dlsym(RTLD_NEXT, "socket");
    realgetpeername = dlsym(RTLD_NEXT, "getpeername");
    realgetsockname = dlsym(RTLD_NEXT, "getsockname");
#else
    lib = dlopen(LIBCONNECT, RTLD_LAZY);
    realconnect = dlsym(lib, "connect");
    realsocket = dlsym(lib, "socket");
    realgetpeername = dlsym(lib, "getpeername");
    realgetsockname = dlsym(lib, "getsockname");
    dlclose(lib);
#endif
    inet_pton(AF_INET6, "::ffff:0.0.0.0", &ipv4mapped);

    get_environment();
}

static int get_environment()
{
    static int done = 0;
    int loglevel = MSGERR;
    char *logfilename = NULL;
    char *env;

    if (done)
        return (0);

    /* Determine the logging level */
#ifndef ALLOW_MSG_OUTPUT
    set_log_options(MSGNONE, NULL, 0);
#else
    if ((env = getenv("TNAT64_DEBUG")))
        loglevel = atoi(env);
    if (((env = getenv("TNAT64_DEBUG_FILE"))) && !suid)
        logfilename = env;
    set_log_options(loglevel, logfilename, 1);
#endif

    done = 1;

    return (0);
}

static int get_config()
{
    static int done = 0;

    if (done)
        return (0);

    /* Determine the location of the config file */
#ifdef ALLOW_ENV_CONFIG
    if (!suid)
        conffile = getenv("TNAT64_CONF_FILE");
#endif

    /* Read in the config file */
    config = malloc(sizeof(*config));
    if (!config)
        return (0);
    read_config(conffile, config);
    if (config->paths)
        show_msg(MSGDEBUG, "First lineno for first path is %d\n", config->paths->lineno);

    done = 1;

    return (0);

}

int socket(SOCKET_SIGNATURE)
{
    /* If the real socket doesn't exist, we're stuffed */
    if (realsocket == NULL)
    {
        show_msg(MSGERR, "Unresolved symbol: socket\n");
        return (-1);
    }
    if ((__domain == AF_INET) && ((__type & SOCK_TYPE_MASK) == SOCK_STREAM))
    {
        return realsocket(AF_INET6, __type, __protocol);
    }
    else
    {
        return realsocket(__domain, __type, __protocol);
    }
}

int connect(CONNECT_SIGNATURE)
{
    struct sockaddr_in *connaddr;
    char addrbuffer[INET6_ADDRSTRLEN];
    struct sockaddr_in6 dest_address6;
    int sock_type = -1;
    socklen_t sock_type_len = sizeof(sock_type);
    struct prefixent *path;
    int failed = 0;

    /* If the real connect doesn't exist, we're stuffed */
    if (realconnect == NULL)
    {
        show_msg(MSGERR, "Unresolved symbol: connect\n");
        return (-1);
    }

    show_msg(MSGDEBUG, "Got connection request\n");

    connaddr = (struct sockaddr_in *)__addr;

    /* Get the type of the socket */
    getsockopt(__fd, SOL_SOCKET, SO_TYPE, (void *)&sock_type, &sock_type_len);

    /* If this isn't an INET socket for a TCP stream we can't  */
    /* handle it, just call the real connect now               */
    if ((connaddr->sin_family != AF_INET) || (sock_type != SOCK_STREAM))
    {
        show_msg(MSGDEBUG, "Connection isn't a TCP/IPv4 stream, ignoring\n");
        return realconnect(__fd, __addr, __len);
    }

    /* If we haven't initialized yet, do it now */
    get_config();

    show_msg(MSGDEBUG, "Got connection request for socket %d to " "%s:%d\n", __fd, inet_ntoa(connaddr->sin_addr), connaddr->sin_port);

    /* If the address is local call realconnect */
    if (!(is_local(config, &(connaddr->sin_addr))))
    {
        show_msg(MSGDEBUG, "Connection for socket %d is local\n", __fd);
        return realconnect(__fd, __addr, __len);
    }

    /* Don't retry more than once */
    while (failed < 2)
    {
        if (current_af == AF_INET)
        {
            /* Construct the IPv4-mapped IPv6 address */
            dest_address6.sin6_family = AF_INET6;
            dest_address6.sin6_port = connaddr->sin_port;
            dest_address6.sin6_flowinfo = 0;
            dest_address6.sin6_scope_id = 0;
            memcpy(&dest_address6.sin6_addr, &ipv4mapped, sizeof(struct in6_addr));
            memcpy(&dest_address6.sin6_addr.s6_addr[NAT64PREFIXLEN], &connaddr->sin_addr, sizeof(struct in_addr));
            if (inet_ntop(AF_INET6, &dest_address6.sin6_addr, addrbuffer, sizeof(addrbuffer)))
            {
                show_msg(MSGDEBUG, "Trying IPv4-mapped IPv6 address %s...\n", addrbuffer);
            }

            if (realconnect(__fd, (struct sockaddr *)&dest_address6, sizeof(struct sockaddr_in6)) == 0)
            {
                show_msg(MSGDEBUG, "Success.\n");
                return 0;
            }
            if (errno != ENETUNREACH)
            {
                show_msg(MSGDEBUG, "Error: %d (%s)\n", errno, strerror(errno));
                return -1;
            }
            else
            {
                show_msg(MSGDEBUG, "Error: %d (%s)\n", errno, strerror(errno));
                current_af = AF_INET6;
                failed++;
            }
        }
        else
        {
            /* Ok, so its not local, we need a path to the net */
            pick_prefix(config, &path, &(connaddr->sin_addr), ntohs(connaddr->sin_port));

            show_msg(MSGDEBUG, "Picked prefix %s for connection\n", (path->address ? path->address : "(Not Provided)"));
            if (path->address == NULL)
            {
                if (path == &(config->defaultprefix))
                    show_msg(MSGERR, "Connection needs to be made " "via default prefix but " "the default prefix has not " "been specified\n");
                else
                    show_msg(MSGERR, "Connection needs to be made " "via path specified at line " "%d in configuration file but " "the prefix has not been " "specified for this path\n", path->lineno);

            }
            else
            {
                /* Construct the NAT64-ed address */
                dest_address6.sin6_family = AF_INET6;
                dest_address6.sin6_port = connaddr->sin_port;
                dest_address6.sin6_flowinfo = 0;
                dest_address6.sin6_scope_id = 0;
                memcpy(&dest_address6.sin6_addr, &path->prefix, sizeof(struct in6_addr));
                memcpy(&dest_address6.sin6_addr.s6_addr[12], &connaddr->sin_addr, sizeof(struct in_addr));
                if (inet_ntop(AF_INET6, &dest_address6.sin6_addr, addrbuffer, sizeof(addrbuffer)))
                {
                    show_msg(MSGDEBUG, "Trying IPv6 address %s...\n", addrbuffer);
                }

                if (realconnect(__fd, (struct sockaddr *)&dest_address6, sizeof(struct sockaddr_in6)) == 0)
                {
                    return 0;
                }
                if (errno != ENETUNREACH)
                {
                    return -1;
                }
                else
                {
                    current_af = AF_INET;
                    failed++;
                }
            }


        }
    }


    /* If we haven't found a valid NAT64 prefix to route the connection, we return ECONNREFUSED */
    errno = ECONNREFUSED;
    return -1;
}

static char afs[][16] = {
    "AF_UNSPEC",
    "AF_UNIX",
    "AF_INET",
    "AF_AX25",
    "AF_IPX",
    "AF_APPLETALK",
    "AF_NETROM",
    "AF_BRIDGE",
    "AF_ATMPVC",
    "AF_X25",
    "AF_INET6"
};

int getpeername(GETPEERNAME_SIGNATURE)
{
    /* If the real getpeername doesn't exist, we're stuffed */
    if (realgetpeername == NULL)
    {
        show_msg(MSGERR, "Unresolved symbol: getpeername\n");
        return (-1);
    }

    /* If we haven't initialized yet, do it now */
    get_config();

    show_msg(MSGDEBUG, "Got getpeername call for socket %d\n", __fd);
    struct sockaddr_in6 realpeer;
    socklen_t needlen = *__len;
    socklen_t realpeerlen = sizeof(realpeer);
    int ret = realgetpeername(__fd, __addr, &needlen);
    struct sockaddr_in * result;
    if (*__len < sizeof(struct sockaddr_in))
    {
        *__len = sizeof(struct sockaddr_in);
        errno = EINVAL;
        return -1;
    }
    if (__addr->sa_family <= 10)
        show_msg(MSGDEBUG, "Address family is %s\n", afs[__addr->sa_family]);
    if (__addr->sa_family == AF_INET6)
    {
        int ret = realgetpeername(__fd, (struct sockaddr *)&realpeer, &realpeerlen);
        if ((!memcmp(&realpeer.sin6_addr, &ipv4mapped, NAT64PREFIXLEN)) || (check_prefix(config, &realpeer.sin6_addr)))
        {
            result = (struct sockaddr_in *)__addr;
            result->sin_family = AF_INET;
            result->sin_port = realpeer.sin6_port;
            memcpy(&result->sin_addr, &realpeer.sin6_addr.s6_addr[12], sizeof(struct in_addr));
            *__len = sizeof(struct sockaddr_in);
            return ret;
        }
    }
return ret;
}

int getsockname(GETSOCKNAME_SIGNATURE)
{
    /* If the real getsockname doesn't exist, we're stuffed */
    if (realgetsockname == NULL)
    {
        show_msg(MSGERR, "Unresolved symbol: getsockname\n");
        return (-1);
    }
    if (realgetpeername == NULL)
    {
        show_msg(MSGERR, "Unresolved symbol: getpeername\n");
        return (-1);
    }

    /* If we haven't initialized yet, do it now */
    get_config();

    show_msg(MSGDEBUG, "Got getsockname call for socket %d\n", __fd);
    struct sockaddr_in6 realpeer;
    socklen_t needlen = *__len;
    socklen_t realpeerlen = sizeof(realpeer);
    int ret = realgetsockname(__fd, __addr, &needlen);
    struct sockaddr_in * result;
    if (*__len < sizeof(struct sockaddr_in))
    {
        *__len = sizeof(struct sockaddr_in);
        errno = EINVAL;
        return -1;
    }
    if (__addr->sa_family <= 10)
        show_msg(MSGDEBUG, "Address family is %s\n", afs[__addr->sa_family]);
    if (__addr->sa_family == AF_INET6)
    {
        int ret = realgetpeername(__fd, (struct sockaddr *)&realpeer, &realpeerlen);
        if ((!memcmp(&realpeer.sin6_addr, &ipv4mapped, NAT64PREFIXLEN)) || (check_prefix(config, &realpeer.sin6_addr)))
        {
            result = (struct sockaddr_in *)__addr;
            result->sin_family = AF_INET;
            result->sin_port = 0;
            memset(&result->sin_addr, 0, sizeof(struct in_addr));
            *__len = sizeof(struct sockaddr_in);
            return ret;
        }
    }
    return ret;
}

